File: bookstore-py/bookstore/handler/review.py
import json
from http.server import BaseHTTPRequestHandler

from bookstore.handler.response import write_error, write_json
from bookstore.store.review import ReviewStore


class ReviewHandler:
    def __init__(self, reviews: ReviewStore):
        self.reviews = reviews

    def list_reviews(self, h: BaseHTTPRequestHandler, params: dict[str, str], query: dict[str, str]) -> None:
        try:
            book_id = int(params["id"])
        except (KeyError, ValueError):
            write_error(h, 500, "invalid book id")
            return
        try:
            reviews = self.reviews.list_by_book(book_id)
        except Exception:
            write_error(h, 500, "failed to list reviews")
            return
        write_json(h, 200, reviews)

    def create_review(self, h: BaseHTTPRequestHandler, params: dict[str, str], query: dict[str, str]) -> None:
        try:
            book_id = int(params["id"])
        except (KeyError, ValueError):
            write_error(h, 500, "invalid book id")
            return
        length = int(h.headers.get("Content-Length") or "0")
        raw = h.rfile.read(length) if length else b""
        try:
            body = json.loads(raw or b"{}")
        except json.JSONDecodeError:
            write_error(h, 400, "invalid body")
            return
        try:
            review = self.reviews.create(
                book_id,
                int(body.get("rating", 0)),
                body.get("review_text", ""),
            )
        except Exception as e:
            write_error(h, 500, str(e))
            return
        write_json(h, 201, review)

File: src/handler/review.ts
import type { ReviewStore } from "../store/review";
import { writeError, writeJSON } from "./response";

export class ReviewHandler {
  constructor(private reviews: ReviewStore) {}

  listReviews = (req: Request): Response => {
    const bookID = parseInt((req as any).params?.id);
    if (isNaN(bookID)) return writeError(500, "invalid book id");
    try {
      const reviews = this.reviews.listByBook(bookID);
      return writeJSON(200, reviews);
    } catch {
      return writeError(500, "failed to list reviews");
    }
  };

  createReview = async (req: Request): Promise<Response> => {
    const bookID = parseInt((req as any).params?.id);
    if (isNaN(bookID)) return writeError(500, "invalid book id");
    let body: { rating: number; review_text: string };
    try {
      body = await req.json();
    } catch {
      return writeError(400, "invalid body");
    }
    try {
      const review = this.reviews.create(bookID, body.rating, body.review_text);
      return writeJSON(201, review);
    } catch (e: any) {
      return writeError(500, e?.message ?? "failed to create review");
    }
  };
}
